src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,320 @@
+            streamSize = tb.getInt(2, i);
+          } else if (col2Type == Type.LONG_TYPE) {
+            streamSize = (int) tb.getLong(2, i);
+          } else {
+            throw new DbException("StreamSize must be of type INT or LONG");
+          }
+          Preconditions.checkState(partitionSize >= 0,
+              "Worker cannot have a negative StreamSize: %d", streamSize);
+        }
+        streamCounts.set(workerID - 1, streamSize);
+      }
+    }
+    // Convert samplePct to sampleSize if using a percentage sample.
+    if (isPercentageSample) {
+      sampleSize = Math.round(totalTupleCount * (samplePercentage / 100));
+    }
+    Preconditions.checkState(sampleSize >= 0 && sampleSize <= totalTupleCount,
+        "Cannot extract %s samples from a population of size %s", sampleSize,
+        totalTupleCount);
+
+    // Generate a random distribution across the workers.
+    int[] sampleCounts;
+    if (sampleType == SamplingType.WR) {
+      sampleCounts = withReplacementDistribution(tupleCounts, sampleSize);
+    } else if (sampleType == SamplingType.WoR) {
+      sampleCounts = withoutReplacementDistribution(tupleCounts, sampleSize);
+    } else {
+      throw new DbException("Invalid sampleType: " + sampleType);
+    }
+
+    // Build and return a TupleBatch with the distribution.
+    IntColumnBuilder wIdCol = new IntColumnBuilder();
+    IntColumnBuilder streamSizeCol = new IntColumnBuilder();
+    IntColumnBuilder sampCountCol = new IntColumnBuilder();
+    StringColumnBuilder sampTypeCol = new StringColumnBuilder();
+    for (int i = 0; i < streamCounts.size(); i++) {
+      wIdCol.appendInt(i + 1);
+      streamSizeCol.appendInt(streamCounts.get(i));
+      sampCountCol.appendInt(sampleCounts[i]);
+      sampTypeCol.appendString(sampleType.name());
+    }
+    ImmutableList.Builder<Column<?>> columns = ImmutableList.builder();
+    columns.add(wIdCol.build(), streamSizeCol.build(), sampCountCol.build(),
+        sampTypeCol.build());
+    setEOS();
+    return new TupleBatch(SCHEMA, columns.build());
+  }
+
+  /**
+   * Creates a WithReplacement distribution across the workers.
+   * 
+   * @param tupleCounts
+   *          list of how many tuples each worker has.
+   * @param sampleSize
+   *          total number of samples to distribute across the workers.
+   * @return array representing the distribution across the workers.
+   */
+  private int[] withReplacementDistribution(List<Integer> tupleCounts,
+      int sampleSize) {
+    int[] distribution = new int[tupleCounts.size()];
+    int totalTupleCount = 0;
+    for (int val : tupleCounts) {
+      totalTupleCount += val;
+    }
+
+    for (int i = 0; i < sampleSize; i++) {
+      int sampleTupleIdx = rand.nextInt(totalTupleCount);
+      // Assign this tuple to the workerID that holds this sampleTupleIdx.
+      int tupleOffset = 0;
+      for (int j = 0; j < tupleCounts.size(); j++) {
+        if (sampleTupleIdx < tupleCounts.get(j) + tupleOffset) {
```

Comment:
I think this code still works if there are "gaps" in the `workerID` sequence (so some `tupleCounts` entries are 0), but that's pretty fragile code. It would be better to iterate through all the existing `workerID` entries in order (say with a `SortedMap`), and avoid these dummy elements altogether.